### PR TITLE
add Half support for BatchNorm on CPU

### LINKED ISSUE #106543: DISABLED test_batchnorm_nhwc_cpu (__main__.TestNN)
Platforms: linux, mac, macos

This test was disabled because it is failing in CI. See [recent examples](https://hud.pytorch.org/flakytest?name=test_batchnorm_nhwc_cpu&suite=TestNN) and the most recent trunk [workflow logs](https://github.com/pytorch/pytorch/runs/15574080336).

Over the past 3 hours, it has been determined flaky in 6 workflow(s) with 18 failures and 6 successes.

**Debugging instructions (after clicking on the recent samples link):**
DO NOT ASSUME THINGS ARE OKAY IF THE CI IS GREEN. We now shield flaky tests from developers so CI will thus be green but it will be harder to parse the logs.
To find relevant log snippets:
1. Click on the workflow logs linked above
2. Click on the Test step of the job so that it is expanded. Otherwise, the grepping will not work.
3. Grep for `test_batchnorm_nhwc_cpu`
4. There should be several instances run (as flaky tests are rerun in CI) from which you can study the logs.

Test file path: `test_nn.py`

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### PR DESCRIPTION
**Pull request by CaoE**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #102070

Fixes #106543

### Testing

Single core:

shape | fp32 forward / ms | fp16 forward / ms | bf16 forward / ms | fp32 backward / ms | fp16 backward / ms | bf16 backward / ms
-- | -- | -- | -- | -- | -- | --
(1, 4, 256, 256) | 0.7116 | 0.1427 | 0.1744 | 0.2638 | 0.2002 | 0.2556
(1, 32, 100, 100) | 0.8579 | 0.1725 | 0.2077 | 0.3023 | 0.2399 | 0.2995
(32, 16, 200, 200) | 57.3466 | 12.2179 | 13.1320 | 45.9524 | 24.1526 | 24.9882

28 cores:

shape | fp32 forward / ms | fp16 forward / ms | bf16 forward / ms | fp32 backward / ms | fp16 backward / ms | bf16 backward / ms
-- | -- | -- | -- | -- | -- | --
(1, 4, 256, 256) | 0.2571 | 0.0713 | 0.0846 | 0.1140 | 0.0883 |  0.1043
(1, 32, 100, 100) | 0.1077 | 0.0510 | 0.0548 | 0.0700 | 0.0645 | 0.0713
(32, 16, 200, 200) | 5.5060 | 1.4195 | 1.4663 | 6.773 | 3.0886 | 3.1343

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

**Comment by CaoE** [comment]
@ngimel Could you please review this PR ? Thank you.

**Comment by CaoE** [comment]
Hi @mikaylagawarecki  I adjusted tolerance values in for https://github.com/pytorch/pytorch/issues/106543.
I do not have a mac machine to reproduce the issue locally, and temporally rename `test_batchnorm_nhwc_cpu` to test on macos by CI.  May I know how to re-enable `test_batchnorm_nhwc_cpu` ?

**Comment by mikaylagawarecki** [comment]
@CaoE apologies I missed this one, will review tomorrow morning

File: test/test_nn.py [hunk 1]
```diff
@@ -5061,8 +5061,9 @@ def test_hardtanh_backward(self):
         x_grad_ref = torch.where(mask, grad, z)
         self.assertEqual(x.grad, x_grad_ref)
 
-    def test_batchnorm_nhwc_cpu(self):
-        def helper(self, mod, size, dtype, mixed_dtype=False, format=torch.channels_last):
+    # Temporary renaming for CI test
+    def test_batchnorm_nhwc_rename(self):
```
Comment (mikaylagawarecki) [inline test/test_nn.py:5065]: To fix a flaky test issue, I think opening a PR with "Fixes <flaky test issue number>" in the initial commit message would make the test run in CI. In this case, since you went via the renaming method, I think we can rename this back and CI bot will automatically close the flaky test issue after this PR is landed when it detects that the test is no longer flaky

**Review by mikaylagawarecki — APPROVED** [review]
Looks good!

File: test/test_nn.py [hunk 2]
```diff
@@ -5090,20 +5091,25 @@ def helper(self, mod, size, dtype, mixed_dtype=False, format=torch.channels_last
             self.assertTrue(out.is_contiguous(memory_format=format))
             self.assertTrue(ref_out.is_contiguous())
             self.assertEqual(out, ref_out)
-            self.assertEqual(bn.weight.grad, ref_bn.weight.grad)
+            self.assertEqual(bn.weight.grad, ref_bn.weight.grad, atol=precision, rtol=precision)
```
Comment (mikaylagawarecki) [inline test/test_nn.py:5094]: question: in general is it expected that macos needs higher atol/rtol?

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 3]
```diff
@@ -170,9 +173,12 @@ void batch_norm_cpu_channels_last_impl(Tensor& output, const Tensor& input,
 }
 
 template <typename scalar_t>
-void batch_norm_cpu_collect_stats_contiguous_impl(
+typename std::enable_if_t<std::is_same_v<scalar_t, at::opmath_type<scalar_t>>, void>
+batch_norm_cpu_collect_stats_contiguous_impl(
     Tensor& mean, Tensor& var_sum, const Tensor& input) {
 
+  // keep acc_type as opmath_type will use float type when scalar_t==float
```
[the remark this answers is not in the document]
Comment (CaoE) [reply aten/src/ATen/native/cpu/batch_norm_kernel.cpp:180]: The difference for this case is : `acc_type` of float is double and `opmath_type` of float is float.
The reason use `acc_type` is: when use float to compute mean and variance, the tolerence values of BN tests needs to be changed accordingly.
Generally, we may choose `opmath_type` for more uniform code style, but for this case `acc_type` will be better.

File: test/test_nn.py (hunk 2 above)
Comment (CaoE) [reply test/test_nn.py:5094]: For this case, the MPS needs higher atol/rtol for float. I am not familiar with the implementation of MPS, and I guess it may be because the accumulation type is not double for float data type on MPS.
Generally, MPS uses same tolerance value for float data type. As for the float16 data type, based on experience, the comparison between MPS and CPU results may exceed the default tolerance.

File: aten/src/ATen/native/cpu/batch_norm_kernel.cpp [hunk 4]
```diff
@@ -593,37 +608,38 @@ void batch_norm_cpu_backward_channels_last_impl(Tensor& grad_input, Tensor& grad
...
+typename std::enable_if_t<!std::is_same_v<scalar_t, at::opmath_type<scalar_t>>, void>
+batch_norm_cpu_contiguous_impl(Tensor& output, const Tensor& input,
     const Tensor& weight, const Tensor& bias, const Tensor& save_mean, const Tensor& save_invstd,
     const Tensor& running_mean, const Tensor& running_var, bool train, double eps) {
-
-  using bVec = Vectorized<BFloat16>;
-  using fVec = Vectorized<float>;
+  using opmath_t = at::opmath_type<scalar_t>;
+  using bVec = Vectorized<scalar_t>;
+  using fVec = Vectorized<at::opmath_type<scalar_t>>;
```
[the remark this answers is not in the document]
Comment (CaoE) [reply aten/src/ATen/native/cpu/batch_norm_kernel.cpp]: Thanks for the comments. Fixed it.

**Comment by CaoE** [comment]
The unexpected success is same as https://github.com/pytorch/pytorch/pull/77496#issuecomment-1126925808.
Is is due to nnc fuser is not enabled with float16: https://github.com/pytorch/pytorch/blob/main/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L937.

**Comment by mikaylagawarecki** [comment]
@clee2000 hm could I check why there isn't signal on the PR that the test is failing

**Comment by clee2000** [comment]
> @clee2000 hm could I check why there isn't signal on the PR that the test is failing

It runs on mac, which runs on the periodic workflow.  I added the ciflow/periodic tag so it will get run on this PR.  It's a similar situation for the slow tests, where I also added the ciflow/slow tag

**Comment by CaoE** [comment]
@clee2000 https://hud.pytorch.org/pytorch/pytorch/commit/b6a1d3fb97ca8eeccf15a4c495fdd1af4b197f88.
The tolerance has been set to 1e-4, but it seems to have not taken effect:
 Greatest absolute difference: 1.6689300537109375e-05 at index (0,) (up to 1e-05 allowed)
 Greatest relative difference: 3.636979454313405e-05 at index (0,) (up to 1.3e-06 allowed)

**Review by mingfeima — APPROVED** [review]

**Comment by CaoE** [comment]
@clee2000 The failing should be fixed. Could you please import this PR to check if it can be landed to main. Thanks.

**Comment by mikaylagawarecki** [comment]
@CaoE I don't think we need to import the PR to check whether it can be landed, as the ciflow/slow and ciflow/periodic tags that were added run the jobs that were previously failing on the PR
